save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/CloneRepositoryController.kt

```diff
@@ -95,72 +85,63 @@ class CloneRepositoryController(
+        testSuitesFilter: (TestSuite) -> Boolean,
+        testSuitesSourceResolver: (T) -> List<TestSuitesSource>,
     ): Mono<StringResponse> {
-        val project = with(executionRequest.project) {
-            projectService.findByNameAndOrganizationName(name, organization.name)
-        } ?: return Mono.just(ResponseEntity.status(HttpStatus.NOT_FOUND).body("Project doesn't exist"))
-
-        val newExecution = createNewExecution(project, username, executionType, configProperties.initialBatchSize, executionRequest.sdk)
-        log.info("Sending request to preprocessor (executionType $executionType) to start save file for project id=${project.id}")
-        val uri = when (executionType) {
-            ExecutionType.GIT -> "/upload"
-            ExecutionType.STANDARD -> "/uploadBin"
-        }
-        return files.updateExecution(newExecution).flatMap { savedExecution ->
-            preprocessorWebClient.post()
-                .uri(uri)
-                .bodyValue(updateExecutionInRequest(executionRequest, savedExecution))
-                .retrieve()
-                .toEntity()
-        }
-    }
-
-    private fun createNewExecution(
-        project: Project,
-        username: String,
-        type: ExecutionType,
-        batchSize: Int,
-        sdk: Sdk,
-    ): Execution {
-        val execution = Execution.stub(project).apply {
-            status = ExecutionStatus.PENDING
-            this.batchSize = batchSize
-            this.sdk = sdk.toString()
-            this.type = type
-            id = executionService.saveExecutionAndReturnId(this, username)
+        val projectCoordinates = with(executionRequest.project) {
+            ProjectCoordinates(organization.name, name)
         }
-        log.info("Creating a new execution id=${execution.id} for project id=${project.id}")
-        return execution
-    }
-
-    private fun Flux<FileInfo>.updateExecution(
-        execution: Execution,
-    ): Mono<Execution> = map {
-        it.toFileKey()
+        val testSuiteIdsMono = blockingToMono { testSuitesSourceResolver(executionRequest) }
+            .flatMapIterable { it }
+            .map { testSuitesSource ->
+                testSuitesSourceSnapshotStorage.list(testSuitesSource.organization.name, testSuitesSource.name)
+                    .collectList()
+                    .map { keys ->
+                        keys.maxByOrNull(TestSuitesSourceSnapshotKey::creationTimeInMills)?.version
+                            ?: throw IllegalStateException("Failed to detect latest version for $testSuitesSource")
+                    }
+                    .flatMapIterable { testSuitesService.getBySourceAndVersion(testSuitesSource, it) }
+                    .filter(testSuitesFilter)
+                    .map { it.requiredId() }
+            }
+            .let {
+                Flux.concat(it)
+            }
+            .collectList()
+        return fileStorage.convertToLatestFileInfo(projectCoordinates, shortFiles)
```

Comment:
Is it a good idea to write reactive methods like this? Isn't reactivity broken here?

Reply:
I didn't get where is a problem. Can change it to extension function for `Flux<ShortFileInfo>` to make it more 'reactive' style

Reply:
added `uploadedMillis` to `ShortFileInfo` and remove `convertToLatestFileInfo`